core-bundle/src/EventListener/SubrequestCacheListener.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\HttpKernel\HttpCache\ResponseCacheStrategy;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+class SubrequestCacheListener
+{
+    public const MERGE_CACHE_HEADER = 'Sf-Merge-Cache-Control';
+
+    /**
+     * @var ResponseCacheStrategy[]
+     */
+    private $strategyStack = [];
+
+    /**
+     * @var ResponseCacheStrategy|null
+     */
+    private $currentStrategy;
+
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        if (KernelInterface::MASTER_REQUEST !== $event->getRequestType()) {
+            return;
+        }
+
+        if ($this->currentStrategy) {
+            $this->strategyStack[] = $this->currentStrategy;
+        }
+
+        $this->currentStrategy = new ResponseCacheStrategy();
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        $response = $event->getResponse();
+
+        if ($this->currentStrategy && $response->headers->has(self::MERGE_CACHE_HEADER)) {
+            if (KernelInterface::MASTER_REQUEST === $event->getRequestType()) {
```

Comment:
We could store the result in an `$isMasterRequest` variable and reuse it on line 59. WDYT?

Reply:
changed in e50d1d864674632ced16783d6731d4f37832cf00